addons/account/models/account_invoice.py

```diff
@@ -949,6 +949,13 @@ def _track_subtype(self, init_values):
             return 'account.mt_invoice_created'
         return super(AccountInvoice, self)._track_subtype(init_values)
 
+    @api.multi
+    def _get_groups(self):
+        self.ensure_one()
+        groups = self.env['account.tax.group']
+        for line in self.tax_line_ids:
+            groups |= line.tax_id.tax_group_id
+        return groups
```

Comment:
From what I can see you only need the _get_groups and get_lines_in_group for the report, correct me if I'm wrong but in the first method you iterate on all lines and return their tax_group_id, and in the second method you iterate on those group to get back the lines because you need the tax_group_id name and the amount_in_group that is set on the line in the report is that correct? You also sort all of that by the sequence on tax_group_id.

If it is the case, here is what I would have do, first of all I would have sorted the recordset in _get_groups using "returns groups.sorted()" which will remove duplicate and sort them using the order set on account.tax.group (which is the sequence). The other thing is that I would have written a method which would return a list of tuple containing [(tax_group_id_name, line_amount_in_group), (tax_group_id_name2, line_amount_in_group2),... ]
which is more efficient and in the report you just have to call the method, iterate over each element of the list and you have access in one call to all the information you need. It will also be clearer for someone reading the code to understand what's happening.